db/src/db.rs

```diff
@@ -995,6 +1009,57 @@ pub fn update_partition_map(conn: &rusqlite::Connection, partition_map: &Partiti
     Ok(())
 }
 
+fn metadata_assertions(conn: &rusqlite::Connection, tx_id: Entid, comitted: bool) -> Result<Vec<(Entid, Entid, TypedValue, bool)>> {
+    let (params, sql_stmt) = match comitted {
+        true => {
+            (
+                vec![&tx_id as &ToSql],
+                format!(r#"
+                    SELECT e, a, v, value_type_tag, added
+                    FROM transactions
+                    WHERE tx = ? AND a IN {}
+                    ORDER BY e, a, v, value_type_tag, added"#,
+                    entids::METADATA_SQL_LIST.as_str()
+                )
+            )
+        },
+        false => {
+            (
+                vec![],
+                format!(r#"
+                    SELECT e, a, v, value_type_tag, added FROM
+                    (
+                        SELECT e0 as e, a0 as a, v0 as v, value_type_tag0 as value_type_tag, 1 as added
+                        FROM temp.search_results
+                        WHERE a0 IN {} AND added0 IS 1 AND ((rid IS NULL) OR
+                            ((rid IS NOT NULL) AND (v0 IS NOT v)))
+
+                        UNION
+
+                        SELECT e0 as e, a0 as a, v, value_type_tag0 as value_type_tag, 0 as added
+                        FROM temp.search_results
+                        WHERE a0 in {} AND rid IS NOT NULL AND
+                        ((added0 IS 0) OR
+                            (added0 IS 1 AND search_type IS ':db.cardinality/one' AND v0 IS NOT v))
+                    
+                    ) ORDER BY e, a, v, value_type_tag, added"#,
+                    entids::METADATA_SQL_LIST.as_str(), entids::METADATA_SQL_LIST.as_str()
+                )
+            )
+        }
+    };
+    
+    // TODO: use concat! to avoid creating String instances.
+    let mut stmt = conn.prepare_cached(&sql_stmt)?;
+    let m: Result<Vec<_>> = stmt.query_and_then(&params[..], |row| -> Result<(Entid, Entid, TypedValue, bool)> {
```

Comment:
Let's extract this to a helper function -- `db::row_to_transaction_assertion`.  Maybe add `db::row_to_datom` as well (without `added`).